-DUSE_XSIMD doesn't work Windows 10 (or Linux) on a certain variable, numpy_expr.hpp(736,53): error: no type named 'simd_iterator'

This one has thrown me off a bit, it won't compile with the -DUSE_XSIMD flag.  The line that gives me issues is the one starting with `random_component` - if I comment that line out (and remove it from the following line), there are no errors.  Here's a minimal example - note this is Windows 10, using either MKL or OpenBLAS, Pythran 0.11.0:

    #pythran export generate_paths(int, int, int, int, float64, float64, float64[:,:] order(C), float64[:,:,:] order(C))
    import numpy as np
    def generate_paths(N_assets, sims, Simulated_time_steps, timesteps, timestep, drift, CurrentVol, randnums3Dcorr):
        
        Simulated_paths = np.zeros((N_assets, sims, Simulated_time_steps))
            
        for i in range(Simulated_time_steps):
            
            randnumscorr = randnums3Dcorr[:,:, i]
            
            dt = np.array((timestep, timestep*2, timestep*3))          
            drift_component = np.multiply( drift - 0.5*CurrentVol**2, dt.reshape(dt.shape[0],1))
            random_component = np.multiply(randnumscorr.T , np.multiply(CurrentVol.reshape(CurrentVol.shape[0],) ,  np.sqrt(dt) ))
            Simulated_paths[:,:,i] = np.exp( drift_component.reshape(drift_component.shape[0],1) + random_component.T)
        
        return Simulated_paths
    
    # [NOTE] uncomment the below if you want to run it in Python...
    #

    # if __name__ == "__main__":
        
    #     N_assets = 3
    #     sims = 8192
    #     Simulated_time_steps = 20
    #     timesteps = 20
    #     timestep = 1/365
    #     drift = 0.0
    #     CurrentVol = np.array([0.5,0.4,0.3])
    #     CurrentVol = CurrentVol.reshape(CurrentVol.shape[0],1)
    #     randnums3Dcorr = np.random.rand(N_assets,sims,timesteps)
        
    #     Simulated_paths = generate_paths(N_assets, sims, Simulated_time_steps, timesteps, timestep, drift, CurrentVol, randnums3Dcorr)

The main error is:

`pythran\pythonic/include/types/numpy_expr.hpp(736,53): error: no type
      named 'simd_iterator'` in '(anonymous namespace)::pythonic::types::numpy_texpr<(anonymous
      namespace)::pythonic::types::broadcasted<(anonymous namespace)::pythonic::types::numpy_expr<(anonymous
      namespace)::pythonic::numpy::functor::multiply, (anonymous namespace)::pythonic::types::ndarray<double, (anonymous
      namespace)::pythonic::types::pshape<long>>, (anonymous namespace)::pythonic::types::numpy_expr<(anonymous
      namespace)::pythonic::numpy::functor::sqrt, (anonymous namespace)::pythonic::types::ndarray<double, (anonymous
      namespace)::pythonic::types::pshape<std::integral_constant<long, 3>>>>>>>'
        `typename std::remove_reference<Args>::type::simd_iterator...>;`

Although there are quite a few more, but that appears to be the source.  Not sure what's going on here.  I tried rewriting the line `random_component` probably 10 different ways, removing all broadcasting with repeat statements, and moving the variables around that comprise that line (moving the assignments to separate variables on separate lines).  Still always throws the same error.  I'm completely at a loss why this particular line is causing the -DUSE_XSIMD to fail.  I opened up `numpy_expr.hpp` and it has declarations for the XSIMD options, and includes `nditerator.hpp` which includes `xsimd.hpp` of which I have 8.0.5 installed from conda-forge which sits in the pkgs directory.  I've tried compiling in an OpenBLAS environment but it has the same issue.

That isn't the actual code because the real code is far too big to troubleshoot.  Plus the difference isn't much, and the same line errors out anyhow.  If the actual code is needed I can supply that, although it appears with my limited knowledge fixing the example will also fix my code...  The only difference is highlighted below, which I'm guessing is just the underlying type inferred by Pythran:

pythran\pythonic/include/types/numpy_expr.hpp(736,53): error: no type named
      'simd_iterator' in '(anonymous namespace)::pythonic::types::numpy_texpr<(anonymous
      namespace)::pythonic::types::broadcasted<(anonymous namespace)::pythonic::types::numpy_expr<(anonymous
      namespace)::pythonic::numpy::functor::multiply, (anonymous namespace)::pythonic::types::ndarray<double, (anonymous
      namespace)::pythonic::types::pshape<long>>, (anonymous namespace)::pythonic::types::numpy_expr<(anonymous
      namespace)::pythonic::numpy::functor::sqrt, (anonymous namespace)::pythonic::types::ndarray<double, (anonymous
      `namespace)::pythonic::types::array_base<long, 1, (anonymous namespace)::pythonic::types::tuple_version`>>>>>>'
        typename std::remove_reference<Args>::type::simd_iterator...>;

NOTE:  I also tried using the current Pythran archive from GitHub as of 3/8/2022 with a: python setup.py install and the problems still are there.  Also tried the newest XSIMD libraries.  So it seems to be either a bug or a known issue with how I structured the variable inside the loop w/ XSIMD.  Here's the full error log as well as the generated .cpp file and .py file: [xsimd_errors_files.zip](https://github.com/serge-sans-paille/pythran/files/8224349/xsimd_errors_files.zip)

I setup my WSL Linux Debian environment and it gives the same error, so it's not platform dependent - looks like I found a bug.  This one is using gcc 9.3.0 and the pythran-openblas package:

/home/user/anaconda3/lib/python3.8/site-packages/pythran/pythonic/include/types/numpy_expr.hpp:730:11: error: no type named 'simd_iterator' in 'std::remove_reference<{anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::broadcasted<{anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::multiply, {anonymous}::pythonic::types::ndarray<double, {anonymous}::pythonic::types::pshape<long int> >, {anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::sqrt, {anonymous}::pythonic::types::ndarray<double, {anonymous}::pythonic::types::pshape<std::integral_constant<long int, 3> > > > > > > >::type' {aka 'struct {anonymous}::pythonic::types::numpy_texpr<{anonymous}::pythonic::types::broadcasted<{anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::multiply, {anonymous}::pythonic::types::ndarray<double, {anonymous}::pythonic::types::pshape<long int> >, {anonymous}::pythonic::types::numpy_expr<{anonymous}::pythonic::numpy::functor::sqrt, {anonymous}::pythonic::types::ndarray<double, {anonymous}::pythonic::types::pshape<std::integral_constant<long int, 3> > > > > > >'}
  730 |     using simd_iterator = numpy_expr_simd_iterator<
      |           ^~~~~~~~~~~~~

--- a/pythran/pythonic/include/types/numpy_texpr.hpp
+++ b/pythran/pythonic/include/types/numpy_texpr.hpp
@@ -303,8 +303,24 @@ namespace types
     using shape_t = types::array<long, value>;
     using iterator = nditerator<numpy_texpr<broadcasted<E>>>;
     using const_iterator = const_nditerator<numpy_texpr<broadcasted<E>>>;
+    // FIXME: I've got the feeling that this could be improved
     static constexpr bool is_vectorizable = false;
     static constexpr bool is_strided = true;
+
+#ifdef USE_XSIMD
+    using simd_iterator = const_simd_nditerator<numpy_texpr<broadcasted<E>>>;
+    using simd_iterator_nobroadcast = simd_iterator;
+    template <class vectorizer>
+    simd_iterator vbegin(vectorizer) const
+    {
+      return {*this}; // not vectorized anyway
+    }
+    template <class vectorizer>
+    simd_iterator vend(vectorizer) const
+    {
+      return {*this}; // not vectorized anyway
+    }
+#endif
 
     broadcasted<E> arg;
     numpy_texpr() = default;

--- a/pythran/pythonic/types/numpy_texpr.hpp
+++ b/pythran/pythonic/types/numpy_texpr.hpp
@@ -91,7 +91,7 @@ namespace types
   typename numpy_texpr_2<E>::simd_iterator
       numpy_texpr_2<E>::vend(vectorizer) const
   {
-    return {*this}; // ! vectorizable anyway
+    return {*this}; // not vectorizable anyway
   }
 #endif
 

